url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Decoder.java

```diff
@@ -0,0 +1,192 @@
+// JDK6    @Override
+    public void set_msg_sink (IMsgSink msg_sink_) 
+    {
+        msg_sink = msg_sink_;
+    }
+
+    @Override
+    protected boolean next() {
+        switch(state()) {
+        case one_byte_size_ready:
+            return one_byte_size_ready ();
+        case eight_byte_size_ready:
+            return eight_byte_size_ready ();
+        case flags_ready:
+            return flags_ready ();
+        case message_ready:
+            return message_ready ();
+        default:
+            return false;
+        }
+    }
+
+
+
+    private boolean one_byte_size_ready() {
+        //  First byte of size is read. If it is 0xff(-1 for java byte) read 8-byte size.
+        //  Otherwise allocate the buffer for message data and read the
+        //  message data into it.
+        byte first = tmpbuf[0];
+        if (first == -1) {
+            next_step (tmpbuf, 8, eight_byte_size_ready);
+        } else {
+
+            //  There has to be at least one byte (the flags) in the message).
+            if (first == 0) {
+                decoding_error ();
+                return false;
+            }
+            
+            int size = (int) first;
+            if (size < 0) {
+                size = (0xFF) & first;
+            }
+
+            //  in_progress is initialised at this point so in theory we should
+            //  close it before calling zmq_msg_init_size, however, it's a 0-byte
+            //  message and thus we can treat it as uninitialised...
+            if (maxmsgsize >= 0 && (long) (size - 1) > maxmsgsize) {
+                decoding_error ();
+                return false;
+
+            }
+            else {
+                in_progress = new Msg(size-1);
+            }
+
+            next_step (tmpbuf, 1, flags_ready);
+        }
+        return true;
+
+    }
+    
+    private boolean eight_byte_size_ready() {
+        //  8-byte payload length is read. Allocate the buffer
+        //  for message body and read the message data into it.
+        final long payload_length = ByteBuffer.wrap(tmpbuf).getLong();
+
+        //  There has to be at least one byte (the flags) in the message).
+        if (payload_length == 0) {
+            decoding_error ();
+            return false;
+        }
+
+        //  Message size must not exceed the maximum allowed size.
+        if (maxmsgsize >= 0 && payload_length - 1 > maxmsgsize) {
+            decoding_error ();
+            return false;
+        }
+
+        //  Message size must fit within range of size_t data type.
+        if (payload_length - 1 > Integer.MAX_VALUE) {
+            decoding_error ();
+            return false;
+        }
+
+        final int msg_size =  (int)(payload_length - 1);
+        //  in_progress is initialized at this point so in theory we should
+        //  close it before calling init_size, however, it's a 0-byte
+        //  message and thus we can treat it as uninitialized...
+        in_progress = new Msg(msg_size);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code assigns a resource to a class field and there is no method that closes or returns this resource. This can result in a resource leak that slows down or crashes your system.

**Fix**
Create a method in this class that closes the following resource: *in\_progress*. After you use this resource, use this method to prevent a resource leak.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).